package/sbp_settings_daemon/sbp_settings_daemon/src/settings.c

```diff
@@ -288,66 +280,73 @@ static void settings_save_callback(u16 sender_id, u8 len, u8 msg[], void *contex
   fclose(f);
 }
 
-static void settings_write_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+static void settings_write_reject(sbp_tx_ctx_t *tx_ctx,
+                                  const char *section,
+                                  const char *name,
+                                  const char *value)
 {
-  (void)sender_id;
-
-  sbp_tx_ctx_t *tx_ctx = (sbp_tx_ctx_t *)context;
-
-  const char *section = NULL, *name = NULL, *value = NULL, *type = NULL;
-  /* Expect to find at least section, name and value */
-  if ((settings_parse(msg, len, &section, &name, &value, &type) >= SETTINGS_TOKENS_VALUE)
-      && settings_lookup(section, name) != NULL) {
-    /* This setting looks good; we'll leave it to the owner to complain if
-     * there's a problem with the value. */
-    return;
+  if (section != NULL && name != NULL) {
+    piksi_log(LOG_ERR, "Setting %s.%s write rejected", section, name);
+  } else {
+    piksi_log(LOG_ERR, "Setting write rejected");
   }
 
-  piksi_log(LOG_ERR, "Setting %s.%s rejected", section, name);
-
   /* Reply with write response rejecting this setting */
   int buflen = 0;
-  u8 buf[BUFSIZE] = {0};
+  char buf[BUFSIZE] = {0};
   buf[buflen++] = SETTINGS_WR_SETTING_REJECTED;
 
-  if (section != NULL) {
-    strncpy(buf, section, BUFSIZE - buflen);
-    buflen += strlen(section) + 1;
+  int res = settings_format(section, name, value, NULL, buf + buflen, BUFSIZE - buflen);
+
+  if (res <= 0) {
+    piksi_log(LOG_WARNING, "Write reject response formatting failed");
+  } else {
+    buflen += res;
   }
 
-  if (name != NULL) {
-    strncpy(buf + buflen, name, BUFSIZE - buflen);
-    buflen += strlen(name) + 1;
+  sbp_tx_send_from(tx_ctx, SBP_MSG_SETTINGS_WRITE_RESP, buflen, buf, SBP_SENDER_ID);
+}
+
+static void settings_write_cb(u16 sender_id, u8 len, u8 msg[], void *ctx)
+{
+  (void)sender_id;
+
+  sbp_tx_ctx_t *tx_ctx = (sbp_tx_ctx_t *)ctx;
+
+  const char *section = NULL, *name = NULL, *value = NULL, *type = NULL;
+  /* Expect to find at least section, name and value */
+  if (settings_parse(msg, len, &section, &name, &value, &type) < SETTINGS_TOKENS_VALUE) {
+    settings_write_reject(tx_ctx, section, name, value);
```

Comment:
seems like this should return `SETTINGS_WR_PARSE_FAILED`. since it wasn't a separate code path before I could understand keeping existing behavior but logically I feel like we might as well update it now that we have the separation of failure modes